### PR TITLE
ssl: functions to access peer cert issue/expire times in network filters

### PR DESCRIPTION
**Pull request by wozz**
*Description*:

Adds functions to access the issue and expiration times of peer certificates, which can be used in network filters.

*Risk Level*: low
*Testing*: unit tests updated
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

File: include/envoy/ssl/connection.h [hunk 1]
```diff
@@ -73,6 +74,18 @@ class Connection {
    *         Returns {} if there is no local certificate, or no SAN field, or no DNS.
    **/
   virtual std::vector<std::string> dnsSansLocalCertificate() const PURE;
+
+  /**
+   * @return SystemTime the time that the peer certificate was issued and should be considered valid
+   *         from. Returns zero value SystemTime if there is no peer certificate.
+   **/
+  virtual SystemTime validFromPeerCertificate() const PURE;
```
Comment (mattklein123) [inline include/envoy/ssl/connection.h]: Thoughts on whether we want to return a system time here and below or something that is guaranteed UTC? Thoughts @PiotrSikora @lizan?

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 2]
```diff
@@ -361,6 +361,22 @@ std::string SslSocket::subjectLocalCertificate() const {
   return Utility::getSubjectFromCertificate(*cert);
 }
 
+SystemTime SslSocket::validFromPeerCertificate() const {
+  bssl::UniquePtr<X509> cert(SSL_get_peer_certificate(ssl_.get()));
+  if (!cert) {
+    return std::chrono::system_clock::from_time_t(0);
```
Comment (mattklein123) [inline source/extensions/transport_sockets/tls/ssl_socket.cc]: Perhaps it would be better to use an `absl:optional` here and below instead of returning 0?

Comment (wozz) [reply source/extensions/transport_sockets/tls/ssl_socket.cc]: makes sense to me. I would assume the caller would use `peerCertificatePresented()` first to check that these functions would return something meaningful

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 3]
```diff
@@ -361,6 +361,22 @@ std::string SslSocket::subjectLocalCertificate() const {
   return Utility::getSubjectFromCertificate(*cert);
 }
 
+absl::optional<SystemTime> SslSocket::validFromPeerCertificate() const {
+  bssl::UniquePtr<X509> cert(SSL_get_peer_certificate(ssl_.get()));
+  if (!cert) {
+    return absl::optional<SystemTime>();
```
Comment (mattklein123) [inline source/extensions/transport_sockets/tls/ssl_socket.cc]: nit: `absl::nullopt`. Also do you have test coverage for this path here and below?

**Comment by wozz** [comment]
It seems like the test failures are actually due to my change... they're all failing on my new test code.  However, it's not failing for me locally.

**Comment by wozz** [comment]
@mattklein123 looks like tests are passing now

**Pull request merged**